### PR TITLE
Make FloatingPoint require that Self.Magnitude == Self

### PR DESCRIPTION
**Pull request by stephentyrone**
We didn't have the where clause to express this constraint at the time that the FloatingPoint protocol was implemented, but we do now. This is not a semantic change to FloatingPoint, which has always bound IEEE-754 arithmetic types, for which this constraint would necessarily hold, but it does effect the type system.

For example, currently the following function does not type check:
~~~~
func foo<T>(x: T) -> T where T: FloatingPoint {
  var r = x.remainder(dividingBy: 1)
  return r.magnitude
}
~~~~
with this change, it compiles correctly.

This is an ABI change, but a very minor one which should not break any currently-functioning code.

### CODE REVIEW

**Comment by stephentyrone** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
You should be able to drop the constraint from `abs` then, right?

**Comment by stephentyrone** [comment]
@jrose-apple Actually I think I can go further and remove the FloatingPoint `abs` entirely. I'll update this shortly.

**Comment by stephentyrone** [comment]
@swift-ci Please test

File: stdlib/public/Platform/tgmath.swift.gyb [hunk 1]
```diff
@@ -14,8 +14,8 @@ import SwiftShims
 
 // Generic functions implementable directly on FloatingPoint.
 @_transparent
-public func fabs<T: FloatingPoint>(_ x: T) -> T
-  where T.Magnitude == T {
+@available(swift, deprecated: 4.2, renamed: "abs")
```
Comment (xwu) [inline stdlib/public/Platform/tgmath.swift.gyb]: Q: why deprecate this specifically?

Comment (stephentyrone) [reply stdlib/public/Platform/tgmath.swift.gyb]: For the same reason that we deprecate all the `tgmath` overlays whose functionality duplicates APIs available on `[Binary]FloatingPoint`. We don't need to carry two names for the same function into the future, and `abs` is the better name for this. `fabs` is a curious artifact of the fact that C doesn't have overloaded functions.

**Comment by stephentyrone** [comment]
@swift-ci please test

**Pull request merged**